node/forest_libp2p/src/service.rs

```diff
@@ -63,57 +83,76 @@ impl Libp2pService {
         loop {
-            match self.swarm.poll() {
-                Ok(Async::Ready(Some(event))) => match event {
+            select! {
+                swarm_event = swarm_stream.next() => match swarm_event {
+                    Some(MergeEvent::Swarm(event)) => match event {
                     MyBehaviourEvent::DiscoveredPeer(peer) => {
-                        libp2p::Swarm::dial(&mut self.swarm, peer);
+                        libp2p::Swarm::dial(&mut swarm_stream.get_mut().get_mut(), peer);
                     }
                     MyBehaviourEvent::ExpiredPeer(_) => {}
                     MyBehaviourEvent::GossipMessage {
                         source,
                         topics,
                         message,
                     } => {
-                        return Ok(Async::Ready(Option::from(NetworkEvent::PubsubMessage {
+                        info!(self.log, "Got a Gossip Message from {:?}", source);
+                        self.pubsub_sender_out.send(NetworkEvent::PubsubMessage {
                             source,
                             topics,
-                            message,
-                        })));
+                            message
+                        }).await;
                     }
+                    }
+                    Some(MergeEvent::Pubsub(_)) => error!(self.log, "Should not happen"),
+                    None => ()
                 },
-                Ok(Async::Ready(None)) => break,
-                Ok(Async::NotReady) => break,
-                _ => break,
-            }
+                pubsub_event = pubsub_stream.next() => match pubsub_event {
+                    Some(MergeEvent::Pubsub(payload)) => {
+                       swarm_stream.get_mut().get_mut().publish(&payload.0, payload.1);
+                    }
+                    Some(MergeEvent::Swarm(_)) => error!(self.log, "Should not happen"),
+                    None => ()
+                }
+            };
         }
-        Ok(Async::NotReady)
     }
-}
 
-/// Events emitted by this Service to be listened by the NetworkService.
-#[derive(Clone)]
-pub enum NetworkEvent {
-    PubsubMessage {
-        source: PeerId,
-        topics: Vec<TopicHash>,
-        message: Vec<u8>,
-    },
+    /// Returns a `Sender` allowing you to send messages over GossipSub
+    pub fn pubsub_sender(&self) -> Sender<(Topic, Vec<u8>)> {
+        self.pubsub_sender_in.clone()
+    }
+
+    /// Returns a `Receiver` to listen to GossipSub messages
+    pub fn pubsub_receiver(&self) -> Receiver<NetworkEvent> {
+        self.pubsub_receiver_out.clone()
+    }
 }
 
+/// Builds the transport stack that LibP2P will communicate over
 pub fn build_transport(local_key: Keypair) -> Boxed<(PeerId, StreamMuxerBox), Error> {
     let transport = libp2p::tcp::TcpConfig::new().nodelay(true);
-    let transport = libp2p::dns::DnsConfig::new(transport);
+    let transport = libp2p::dns::DnsConfig::new(transport).unwrap();
```

Comment:
Can you avoid unwrapping by handling the error?

Reply:
I figured that if the transport cant be built, might as well crash the program on start? Thoughts?

Reply:
I just think it would be better to handle the error anyway instead of panicing. I see the error is handled and boxed below, so I don't see why not here also. 

Also what is the point of setting up the tcp transport when you replace it with the dns transport this line?

Reply:
Cool, I'll do it 👌